traits/observers/tests/test_trait_event_notifier.py

```diff
@@ -0,0 +1,511 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+import io
+import unittest
+from unittest import mock
+import weakref
+
+from traits.observers._exception_handling import (
+    pop_exception_handler,
+    push_exception_handler,
+)
+from traits.observers._exceptions import NotifierNotFound
+from traits.observers._trait_event_notifier import TraitEventNotifier
+
+
+def dispatch_here(function, event):
+    """ Dispatcher that let the function call through."""
+    function(event)
+
+
+def not_prevent_event(event):
+    """ An implementation of prevent_event that does not prevent
+    any event from being propagated.
+    """
+    return False
+
+
+class DummyObservable:
+    """ Dummy implementation of IObservable for testing purposes.
+    """
+
+    def __init__(self):
+        self.notifiers = []
+
+    def _notifiers(self, force_create):
+        return self.notifiers
+
+    def handler(self, event):
+        pass
+
+
+# Dummy target object that is not garbage collected while the tests are run.
+_DUMMY_TARGET = mock.Mock()
+
+
+def create_notifier(**kwargs):
+    """ Convenient function for creating an instance of TraitEventNotifier
+    for testing purposes.
+    """
+    values = dict(
+        handler=mock.Mock(),
+        target=_DUMMY_TARGET,
+        event_factory=mock.Mock(),
+        prevent_event=not_prevent_event,
+        dispatcher=dispatch_here,
+    )
+    values.update(kwargs)
+    return TraitEventNotifier(**values)
+
+
+class TestTraitEventNotifierCall(unittest.TestCase):
+    """ Test calling an instance of TraitEventNotifier. """
+
+    def setUp(self):
+        push_exception_handler(reraise_exceptions=True)
+        self.addCleanup(pop_exception_handler)
+
+    def tearDown(self):
+        pass
+
+    def test_init_and_call(self):
+
+        handler = mock.Mock()
+
+        def event_factory(*args, **kwargs):
```

Comment:
There's a lot of mocks in these tests (which I'm not quite sure if it's a good thing), but then in some cases alternatives are used (e.g. this `event_factory` function or lambda functions in another related PR). I think it would be good to keep consistency in the tests

Reply:
Yes this test and the next test are not consistent in the choice of how a callable is defined, because they are not written at the same time despite by the same person. While this maybe a bit of an eyesore to some, I don't think this actually causes any issues...
Will attempt to make them consistent anyway. But I don't think this is not going to keep being consistent for a long time.

Reply:
> While this maybe a bit of an eyesore to some, I don't think this actually causes any issues...

Just to clarify, I am not saying `mock.Mock` objects are always an equivalent replacement to defining a function in-place. In some situations, the distinctions are important, but those differences are not relevant here.

Reply:
I did the attempt. See what you think.

I also went over the tests to see where else I could make them consistent... and I really could not do much. Here is an example:
```
    def test_prevent_event_is_used(self):
        # Test prevent_event can stop an event from being dispatched.

        def prevent_event(event):
            return True

        handler = mock.Mock()
        notifier = create_notifier(
            handler=handler,
            prevent_event=prevent_event,
        )

        # when
        notifier(a=1, b=2)

        # then
        handler.assert_not_called()
```

If I define `handler` in place, I will do something else to check it is not called. If I make `prevent_event` a mock, I will have to check its call argument to make sure it is indeed called with `(handler, event)`.  I don't think going one way or another in this test gains much. How it is right now seems readable enough.

Reply:
I think it's fine to not be consistent and use mocks if you're checking the arguments - it's clear why the mock is being used. In other cases, when mock isn't actually being used, it makes me wonder why the difference is there, is it significant or not. So I prefer consistency because then I don't have to worry about some nuance behaviours that might be hiding in those differences.

The fix looks good to me, thanks for the changes! Just a note that the same applies to tests in #1007. 

All of this is on a "nice to have" basis so feel free to ignore my comments if you're busy with more important things.

Reply:
Thank you for the explanation. I will keep that in mind.
I am aware that some of this came from #1007 too, in particular where a mock object may hold references to its call arguments, preventing them from being garbage collected. I also went over the other tests in #1007 following your comment. Another example of that inconsistency maybe this:
```
        notifier = create_notifier(
            observer_handler=observer_handler,
            handler=handler,
            target=target,
            event_factory=lambda value: value,
            prevent_event=lambda event: event != "Fire",
        )
``` 
as opposed to this:
```
        def event_factory(value):
            return value

        def prevent_event(event):
            return event != "Fire"

        notifier = create_notifier(
            observer_handler=observer_handler,
            handler=handler,
            target=target,
            event_factory=event_factory,
            prevent_event=prevent_event,
        )
```
Knowing myself, I don't think I will be in any capacity to be able to hunt down all these kind of inconsistencies.  I am sorry about that.